test/scale/scale_tester.go

```diff
@@ -117,68 +134,56 @@ scaleOutTest:
 			Replicas: int32(0),
 		},
 	}
-	_, _ = client.AppsV1().StatefulSets(statefulSet.Namespace.Name).UpdateScale(context.TODO(), statefulSet.Statefulset.Name, newScale, metav1.UpdateOptions{})
 
-	numberOfRemainingPods := totalNumberOfPods
-	ticker.Reset(time.Minute)
-	tickerCount = 0
-	timeout = time.After(time.Duration(*scaleTestTimeout) * time.Minute)
-scaleInTest:
-	for {
-		select {
-		case <-timeout:
-			break scaleInTest
-		case <-ticker.C:
-			tickerCount = tickerCount + 1
-			numberOfRemainingPods = podutil.CountAllPodsWithMatchingLabel(client, namespace, map[string]string{"group": "azuredisk-scale-tester"}, string(v1.PodRunning))
-			if numberOfRemainingPods <= 0 {
-				break scaleInTest
-			}
+	startScaleIn := time.Now()
+	_, _ = client.AppsV1().StatefulSets(statefulSet.Namespace.Name).UpdateScale(context.TODO(), statefulSet.Statefulset.Name, newScale, metav1.UpdateOptions{})
+	numberOfRemainingPods := podutil.WaitForAllPodsWithMatchingLabelToDelete(*scaleTestTimeout, time.Minute, client, namespace)
+	scaleInPodsCompleted := time.Since(startScaleIn).Minutes()
 
-			if tickerCount == 1 {
-				oneMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
-			} else if tickerCount == 2 {
-				twoMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
-			} else if tickerCount == 3 {
-				threeMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
-			} else if tickerCount == 5 {
-				fiveMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
-			} else if tickerCount == 10 {
-				tenMinPodsDeleted = totalNumberOfPods - numberOfRemainingPods
-			}
-		}
+	var wg sync.WaitGroup
+	numberOfRemainingVolumeAtts := numberOfRunningPods * *volMountedToPod
+	scaleInVACompleted := float64(0)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		numberOfRemainingVolumeAtts = volutil.WaitForAllVolumeAttachmentsToDelete(*scaleTestTimeout, 30*time.Second, client)
+		scaleInVACompleted = time.Since(startScaleIn).Minutes()
```

Comment:
np: we don't need to change this in this PR. Instead of waiting for done signal, should we return the timestamp in a channel and wait for that value? That way we wouldn't update a var from the outer scope in the goroutine.

Reply:
Actually, the easier thing would be to call `time.Sleep(controller.DefaultTimeUntilGarbageCollection)` in the goroutine and wait for it before `WaitForAllReplicaAttachmentsToDetach`. We then do `WaitForAllVolumeAttachmentsToDelete` in the main thread while waiting for the goroutine.

Reply:
As Huichan pointed out, the second solution won't be ideal because we have no guarantee that WaitForAllReplicaAttachmentsToDetach is always under 5 min.

Reply:
As discussed with Lana, we use channel to return an integer and a timestamp inside a struct.